### PR TITLE
Update D3 typings

### PR DESCRIPTION
**Pull request by jtlan**
From 837a1a0:

>Caught some typing errors in `Animators`.
No more `D3.QuantitativeScale`, so had to type some things more explicitly.
* `Scales.InterpolatedColor` was the most affected in that regard.

>Most other changes are to parameterize `Selections` -- Basically always `<void>`, or `<any>` if a datum is bound to it.

Also updated SVG-Typewriter, which had to be pulled along to the new version too.

### CODE REVIEW

File: src/animators/baseAnimator.ts [hunk 1]
```diff
@@ -63,7 +63,7 @@ export module Animators {
       return time;
     }
 
-    public animate(selection: any, attrToProjector: AttributeToProjector): D3.Transition.Transition {
+    public animate(selection: d3.Selection<any>, attrToAppliedProjector: AttributeToAppliedProjector) {
```
Comment (aicioara) [inline src/animators/baseAnimator.ts]: Can you tell me a bit about the difference between `AttributeToProjector` and `AttributeToAppliedProjector`

File: src/animators/rectAnimator.ts [hunk 2]
```diff
@@ -19,25 +19,25 @@ export module Animators {
...
+    protected _startMovingProjector(attrToAppliedProjector: AttributeToAppliedProjector) {
       if (this.isVertical === this.isReverse) {
-        return attrToProjector[this._getMovingAttr()];
+        return attrToAppliedProjector[this._getMovingAttr()];
       }
-      var movingAttrProjector = attrToProjector[this._getMovingAttr()];
-      var growingAttrProjector = attrToProjector[this._getGrowingAttr()];
-      return (d: any, i: number, dataset: Dataset) => {
-        return movingAttrProjector(d, i, dataset) + growingAttrProjector(d, i, dataset);
+      var movingAttrAppliedProjector = attrToAppliedProjector[this._getMovingAttr()];
```
Comment (aicioara) [inline src/animators/rectAnimator.ts]: `movingAttrToAppliedProjector`? Missing the `to`

File: src/components/axes/axis.ts [hunk 3]
```diff
@@ -188,7 +188,7 @@ module Plottable {
     }
 
     protected _generateTickMarkAttrHash(isEndTickMark = false) {
-      var tickMarkAttrHash = {
+      var tickMarkAttrHash: { [key: string]: number | ((d: any) => number) } = {
```
Comment (aicioara) [inline src/components/axes/axis.ts]: Is there a way to avoid this union? Shouldn't the second argument be `[key: string]: (d:any) => number`?

File: src/components/axes/numericAxis.ts [hunk 4]
```diff
@@ -89,7 +89,7 @@ export module Axes {
     public renderImmediately() {
       super.renderImmediately();
 
-      var tickLabelAttrHash = {
+      var tickLabelAttrHash: { [key: string]: number | string | ((d: any) => number) } = {
```
Comment (aicioara) [inline src/components/axes/numericAxis.ts]: Same as above

File: src/components/axes/timeAxis.ts [hunk 5]
```diff
@@ -281,7 +281,7 @@ export module Axes {
 
     private _getIntervalLength(config: TimeAxisTierConfiguration) {
       var startDate = this._scale.domain()[0];
-      var d3Interval = Formatters.timeIntervalToD3Time(config.interval);
+      var d3Interval = Formatters.timeIntervalTod3Time(config.interval);
```
Comment (aicioara) [inline src/components/axes/timeAxis.ts]: Capitalize D here for CamelCase

File: src/components/axes/timeAxis.ts [hunk 6]
```diff
@@ -385,7 +385,7 @@ export module Axes {
         labelPos = tickPos;
       }
 
-      var tickLabels = container.selectAll("." + Axis.TICK_LABEL_CLASS).data(labelPos, (d) => d.valueOf());
+      var tickLabels = container.selectAll("." + Axis.TICK_LABEL_CLASS).data(labelPos, (d) => String(d.valueOf()));
```
Comment (aicioara) [inline src/components/axes/timeAxis.ts]: Why was this needed?

File: src/components/axes/timeAxis.ts [hunk 7]
```diff
@@ -503,7 +503,7 @@ export module Axes {
     }
 
     private _hideOverlappingAndCutOffLabels(index: number) {
-      var boundingBox = this._element.select(".bounding-box")[0][0].getBoundingClientRect();
+      var boundingBox = (<Element> this._element.select(".bounding-box")[0][0]).getBoundingClientRect();
```
Comment (aicioara) [inline src/components/axes/timeAxis.ts]: Just a thought: we should consider if we want to enforce a whitespace between casting and variable or not. We have a mixture in the codebase

File: src/components/interpolatedColorLegend.ts [hunk 8]
```diff
@@ -202,11 +202,11 @@ export module Components {
...
-      var boundingBoxAttr = {
-        x: 0,
-        y: padding,
-        width: 0,
-        height: 0
+      var boundingBoxAttr: { [key: string]: number} = {
+        "x": 0,
+        "y": padding,
+        "width": 0,
+        "height": 0
```
Comment (aicioara) [inline src/components/interpolatedColorLegend.ts]: We should be consistent, in general we don't stringify them (at least it did not happen above).

File: src/components/interpolatedColorLegend.ts [hunk 9]
```diff
@@ -202,11 +202,11 @@ export module Components {
       var swatchX: (d: any, i: number) => number;
       var swatchY: (d: any, i: number) => number;
 
-      var boundingBoxAttr = {
-        x: 0,
-        y: padding,
-        width: 0,
-        height: 0
+      var boundingBoxAttr: { [key: string]: number} = {
```
Comment (aicioara) [inline src/components/interpolatedColorLegend.ts]: Space between `number` and `}`

File: src/components/plots/plot.ts [hunk 10]
```diff
@@ -422,15 +422,16 @@ module Plottable {
      *
      * @param {Dataset[]} [datasets] The Datasets to retrieve the Selections for.
      *   If not provided, Selections will be retrieved for all Datasets on the Plot.
-     * @returns {D3.Selection}
+     * @returns {d3.Selection}
      */
-    public getAllSelections(datasets = this.datasets()): D3.Selection {
-      var allSelections: EventTarget[] = [];
+
```
Comment (aicioara) [inline src/components/plots/plot.ts]: Extra whitespace

File: src/scales/linearScale.ts [hunk 11]
```diff
@@ -53,7 +53,7 @@ export module Scales {
     }
 
     protected _niceDomain(domain: number[], count?: number): number[] {
-      return Utils.D3Scale.niceDomain(this._d3Scale, domain, count);
+      return this._d3Scale.copy().domain(domain).nice(count).domain();
```
Comment (aicioara) [inline src/scales/linearScale.ts]: Why remove the D3Scale method?

File: src/scales/timeScale.ts [hunk 12]
```diff
@@ -24,10 +24,10 @@ export module Scales {
     public tickInterval(interval: string, step?: number): Date[] {
       // temporarily creats a time scale from our linear scale into a time scale so we can get access to its api
       var tempScale = d3.time.scale();
-      var d3Interval = Formatters.timeIntervalToD3Time(interval);
+      var d3Interval: d3.time.Interval = Formatters.timeIntervalTod3Time(interval);
```
Comment (aicioara) [inline src/scales/timeScale.ts]: Capitalize D for CamelCase. Is the explicit type d3.time.Interval needed?

File: src/scales/timeScale.ts [hunk 13]
```diff
@@ -80,8 +80,8 @@ export module Scales {
       return this._d3Scale.ticks(QuantitativeScale._DEFAULT_NUM_TICKS);
     }
 
-    protected _niceDomain(domain: Date[], count?: number) {
-      return Utils.D3Scale.niceDomain(this._d3Scale, domain, count);
+    protected _niceDomain(domain: Date[]) {
+      return this._d3Scale.copy().domain(domain).nice().domain();
```
Comment (aicioara) [inline src/scales/timeScale.ts]: Same as above

File: src/utils/formatters.ts [hunk 14]
```diff
@@ -193,7 +193,7 @@ module Plottable {
      * Transforms the Plottable TimeInterval string into a d3 time interval equivalent.
      * If the provided TimeInterval is incorrect, the default is d3.time.year
      */
-    export function timeIntervalToD3Time(timeInterval: string) {
+    export function timeIntervalTod3Time(timeInterval: string) {
```
Comment (aicioara) [inline src/utils/formatters.ts]: Capitalize D for CamelCase

File: test/components/plots/rectanglePlotTests.ts [hunk 15]
```diff
@@ -14,9 +14,9 @@ describe("Plots", () => {
       { x: 4, y: 4, x2: 5, y2: 5 }
     ];
 
-    var VERIFY_CELLS = (cells: D3.Selection) => {
+    var VERIFY_CELLS = (cells: d3.Selection<any>) => {
       assert.strictEqual(cells[0].length, 5);
-      cells.each(function(d: D3.Selection, i: number) {
+      cells.each(function(d: any, i: number) {
```
Comment (aicioara) [inline test/components/plots/rectanglePlotTests.ts]: Is this an `any`?

File: test/core/componentGroupTests.ts [hunk 16]
```diff
@@ -65,7 +65,8 @@ describe("ComponentGroups", () => {
     componentGroup.remove(c1);
     assert.deepEqual(componentGroup.components(), [c0, c2],
     "removing a Component not in the Group does not remove Components from the Group");
-    assert.strictEqual(svg.node().childNodes[0], (<D3.Selection> (<any> c1)._element).node(), "The Component not in the Group stayed put");
+    assert.strictEqual((<Node> svg.node()).childNodes[0], (<d3.Selection<void>> (<any> c1)._element).node(),
```
Comment (aicioara) [inline test/core/componentGroupTests.ts]: is the cast to `d3.Selection<void>` necessary?

File: src/animators/baseAnimator.ts (hunk 1 above)
Comment (jtlan) [inline src/animators/baseAnimator.ts]: an `AppliedProjector` has the `Dataset` rolled into it, so the signature is different:
```Typescript
/**
 * Retrieves scaled datum property.
 */
export type _Projector = (datum: any, index: number, dataset: Dataset) => any;

export type AppliedProjector = (datum: any, index: number) => any;
```

File: src/animators/rectAnimator.ts (hunk 2 above)
Comment (jtlan) [inline src/animators/rectAnimator.ts]: It's not a dictionary, the type of that object is `AppliedProjector`.

File: src/components/axes/axis.ts (hunk 3 above)
Comment (jtlan) [inline src/components/axes/axis.ts]: I could convert the hash values to a constant
```Typescript
() => 0
```
instead of just using `number`, but that complicates the later code, and d3 does accept the raw `number`s. I wanted to modify as little of the actual code logic as possible in this PR.

I should remove the `<any>`s on the following 4 lines though.

File: src/components/axes/timeAxis.ts (hunk 5 above)
Comment (jtlan) [inline src/components/axes/timeAxis.ts]: Good catch, bad find-replace on my part.

File: src/components/axes/timeAxis.ts (hunk 6 above)
Comment (jtlan) [inline src/components/axes/timeAxis.ts]: The keying function (the second argument to `.data()`) [has to return `string`s](https://github.com/mbostock/d3/wiki/Selections#data); previously the return value was implicitly being cast by Javascript.

File: src/components/axes/timeAxis.ts (hunk 7 above)
Comment (jtlan) [inline src/components/axes/timeAxis.ts]: I prefer the whitespace, personally.

File: src/scales/linearScale.ts (hunk 11 above)
Comment (jtlan) [inline src/scales/linearScale.ts]: It depended on the D3 type `QuantitativeScale` existing, but that type no longer exists. Rather than have it take some sort of union type, I thought it made more sense just to clone the implementation, in particular because the signature of the `nice()` method is different on different scales.

File: test/components/plots/rectanglePlotTests.ts (hunk 15 above)
Comment (jtlan) [inline test/components/plots/rectanglePlotTests.ts]: Yes, because the `d` there is the data value, not the `Selection`. The previous author made a mistake.

File: test/core/componentGroupTests.ts (hunk 16 above)
Comment (jtlan) [inline test/core/componentGroupTests.ts]: No, but it provides better typing safety than just using `<any>`.

File: src/components/interpolatedColorLegend.ts (hunk 8 above)
Comment (jtlan) [inline src/components/interpolatedColorLegend.ts]: Good catch, thanks.

File: src/scales/timeScale.ts (hunk 12 above)
Comment (jtlan) [inline src/scales/timeScale.ts]: Nope, not needed.

File: src/animators/rectAnimator.ts (hunk 2 above)
Comment (aicioara) [inline src/animators/rectAnimator.ts]: aw, my bad. However, in that case the `Attr` bit is extra

Comment (jtlan) [inline src/animators/rectAnimator.ts]: I'll change it so it's clearer.

File: src/components/axes/axis.ts (hunk 3 above)
Comment (aicioara) [inline src/components/axes/axis.ts]: Let's keep the union, but isn't the second argument `[key: string]: (d:any) => number`? Like, it's a map from key to functions? I'm missing something here. How does this even compile?

File: test/core/componentGroupTests.ts (hunk 16 above)
Comment (aicioara) [inline test/core/componentGroupTests.ts]: But it's not an `<any>`. You cast the `_element`, which is of type `d3.Selection<void>` anyway

File: src/components/axes/axis.ts (hunk 3 above)
Comment (jtlan) [inline src/components/axes/axis.ts]: It's a map from strings to `number | (d: any) => number` (number or function).

File: test/core/componentGroupTests.ts (hunk 16 above)
Comment (jtlan) [inline test/core/componentGroupTests.ts]: The test code doesn't know that though, because `_element` isn't public. After we `<any>`-cast c1, any fields on it are also `<any>`.

That said, it's probably better just to rewrite this `assert`.

File: src/components/axes/axis.ts (hunk 3 above)
Comment (aicioara) [inline src/components/axes/axis.ts]: my bad, ty
